### PR TITLE
Use voro++ to compute Voronoi diagrams.

### PR DESCRIPTION
**Pull request by bdice**
**This PR replaces #362, which I rewrote to use a git submodule.** PR #362 was too far out of sync with `next` to salvage.

## Description
This PR adds the voro++ library ([homepage](http://math.lbl.gov/voro++/), [GitHub repo](https://github.com/chr1shr/voro)) written by Chris Rycroft (@chr1shr) to freud. This will replace the existing Voronoi methods, which used less sophisticated algorithms and were quite slow.

## Motivation and Context
Issue #305 describes challenges with Voronoi performance -- this continues to be a bottleneck for my current analyses, so I found it worthwhile to rewrite this entire feature using voro++.

## How Has This Been Tested?
I have adapted existing Voronoi tests, which work as expected in both 2D and 3D systems. Some changes (like reordering of vertices in a Voronoi cell) are expected, so I added functions to sort the vertices in a deterministic way for the purpose of testing. Currently, this new approach is about 32x faster for a system of 5000 random points in a cubic box of side length 10.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [x] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

### CODE REVIEW

**Comment by vyasr** [comment]
Good call on making a new PR. This approach also has the ancillary approach of avoiding having our history cluttered with all the Voro++ files. I'll aim to review later today.

File: cpp/locality/Voronoi.cc
Comment (vyasr) [inline cpp/locality/Voronoi.cc:42]: Why do we have 3 blocks in each direction? And what is the logic for the choice of the initial memory allocation?

File: cpp/locality/Voronoi.cc
Comment (vyasr) [inline cpp/locality/Voronoi.cc:47]: I'm personally inclined to avoid using the `query_point` terminology at all in this code, since Voronoi constructions don't really make sense for two distinct sets of points. I could be convinced otherwise, though, if you think it's a helpful mental model. Thoughts?

Comment (vyasr) [reply cpp/locality/Voronoi.cc:47]: I also don't think it's a good enough reason to hold up this PR. So if nothing else major crops up I'm not going to bother asking for a change no matter where we come down.

File: tests/test_locality_Voronoi.py
Comment (vyasr) [inline tests/test_locality_Voronoi.py:8]: Add a comment on the reason for the sort order.

**Review by vyasr — APPROVED** [review]
This looks great! I have a couple of questions I'd like answered before you merge, and one comment I'd like added to the tests, but I'm approving contingent on that happening.

File: cpp/locality/Voronoi.cc
Comment (bdice) [reply cpp/locality/Voronoi.cc:47]: I'm just trying to keep it consistent. I don't know what else I would call it without being _more_ strange, since we've standardized terms elsewhere in the codebase.

File: cpp/locality/Voronoi.cc
Comment (bdice) [reply cpp/locality/Voronoi.cc:42]: I added a comment about this marked as `TODO`.

File: tests/test_locality_Voronoi.py
Comment (bdice) [reply tests/test_locality_Voronoi.py:8]: Done.

**Pull request merged**
